distributed/semaphore.py

```diff
@@ -0,0 +1,259 @@
+                client_has_leases = sum(
+                    valmap(len, self.leases_per_client[dead_client]).values()
+                )
+                if client_has_leases:
+                    self._release_client(dead_client)
+            else:
+                self._validation_running = False
+
+
+class Semaphore:
+    def __init__(self, max_leases=1, name=None, client=None):
+        self.client = client or _get_global_client() or get_worker().client
+        self.id = uuid.uuid4().hex
+        self.name = name or "semaphore-" + uuid.uuid4().hex
+        self.max_leases = max_leases
+
+        # TODO: Good idea to create the resource in the init?
+        if self.client.asynchronous:
+            self._started = self.client.scheduler.semaphore_create(
+                name=self.name, max_leases=max_leases
+            )
+        else:
+            self.client.sync(
+                self.client.scheduler.semaphore_create,
+                name=self.name,
+                max_leases=max_leases,
+            )
+            self._started = True
+
+    def __await__(self):
+        async def create_semaphore():
+            await self._started
+            return self
+
+        return create_semaphore().__await__()
+
+    def acquire(self, timeout=None):
+        """
+        Acquire a semaphore.
+
+        If the internal counter is greater than zero, decrement it by one and return True immediately.
+        If it is zero, wait until a release() is called and return True.
+        """
+        return self.client.sync(
+            self.client.scheduler.semaphore_acquire,
+            name=self.name,
+            timeout=timeout,
+            client=self.client.id,
+            identifier=self.id,
+        )
+
+    def release(self):
+        """
+        Release a semaphore.
+
+        Increment the internal counter by one.
+        """
+
+        """ Release the lock if already acquired """
+        return self.client.sync(
+            self.client.scheduler.semaphore_release,
+            name=self.name,
+            client=self.client.id,
+            identifier=self.id,
+        )
+
+    def __enter__(self):
+        self.acquire()
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.release()
+
+    async def __aenter__(self):
+        await self.acquire()
+        return self
+
+    async def __aexit__(self, *args, **kwargs):
+        await self.release()
+
+    def __getstate__(self):
+        return (self.name, self.client.scheduler.address, self.max_leases)
+
+    def __setstate__(self, state):
+        name, address, max_leases = state
+        try:
+            client = get_client(address)
+        except (AttributeError, AssertionError):
+            client = Client(address, set_as_default=False)
+        self.__init__(name=name, client=client, max_leases=max_leases)
+
+    def close(self):
+        self.client.sync(self.client.scheduler.semaphore_close, name=self.name)
```

Comment:
```suggestion
        return self.client.sync(self.client.scheduler.semaphore_close, name=self.name)
```

We need to return the result here.  If we're in async mode then this will be an `asyncio.Future`, and the caller will need to await it.